### ISSUE TITLE
Building on Windows

### ISSUE THREAD

**Issue by tpoppins** [drfrag666/chocolate-doom#1]
Win 7 x64 Pro SP1
CodeBlocks 17.12
TDM-GCC-64 5.1.0.3
RUDE Doom rev 4603 (2019-12-18) via TortoiseSVN

1) Load chocolate.workspace

2) Building the workspace aborts due to SDL headers not found
`D:\msys64\home\xxx\rude-doom\textscreen\txt_sdl.h|24|fatal error: SDL.h: No such file or directory|`

3) Add "D:\msys64\mingw32\include" and "D:\msys64\mingw32\include\SDL" from local MSYS2-32 install under Compiler's Search Directories tab

4) Linker complains about libs not found:
```
<snip>
D:\msys64\home\admin\rude-doom\src\doomtype.h|36|warning: "strncasecmp" redefined|
C:\Program Files (x86)\CodeBlocks\TDM-GCC-64\x86_64-w64-mingw32\include\string.h|109|note: this is the location of the previous definition|
ld.exe||cannot find -lSDLmain|
ld.exe||cannot find -lSDL|
ld.exe||cannot find -lSDL_mixer|
ld.exe||cannot find -lSDL_net|
ld.exe||cannot find -lz.dll|
ld.exe||cannot find -llibFLAC-8|
ld.exe||cannot find -llibogg-0|
ld.exe||cannot find -llibvorbis-0|
ld.exe||cannot find -llibvorbisfile-3|
||error: ld returned 1 exit status|
||=== Build failed: 10 error(s), 332 warning(s) (0 minute(s), 9 second(s)) ===|
```

5) Add "D:\msys64\mingw32\lib" and "D:\msys64\xxx\admin\rude-doom\externlib" to Linker's Search Dirs tab

6) Linker doesn't recognize the libs, same errors as in point 4.

Tried setting the following execs for the linker without success:
dynamic: g++.exe and then x86_64-w64-mingw32-g++.exe
static: ar.exe and then x86_64-w64-mingw32-ar.exe

Compiling under MSYS2/MinGW fails with different errors, I'll address that in next post.

**Comment by tpoppins** [comment]
Under MSYS2/MinGW

1) make aborts with
```
# make
make: *** No targets specified and no makefile found.  Stop.
```

2) run ./autogen.sh:

```
configure.ac:17: installing 'autotools/compile'
configure.ac:13: installing 'autotools/config.guess'
configure.ac:13: installing 'autotools/config.sub'
configure.ac:139: installing 'autotools/install-sh'
configure.ac:139: installing 'autotools/missing'
opl/Makefile.am: installing 'autotools/depcomp'
configure: loading site script /mingw32/etc/config.site
checking build system type... i686-w64-mingw32
checking host system type... i686-w64-mingw32
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.exe
checking for suffix of executables... .exe
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
checking for ranlib... ranlib
checking for python... true
checking for sdl-config... /mingw32/bin/sdl-config
checking for SDL - version >= 1.1.3... yes
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking for Mix_LoadMUS in -lSDL_mixer... yes
checking for SDLNet_UDP_Send in -lSDL_net... yes
checking for src_new in -lsamplerate... yes
checking for zlibVersion in -lz... yes
checking for png_get_io_ptr in -lpng... yes
checking for log in -lm... yes
checking linux/kd.h usability... no
checking linux/kd.h presence... no
checking for linux/kd.h... no
checking dev/isa/spkrio.h usability... no
checking dev/isa/spkrio.h presence... no
checking for dev/isa/spkrio.h... no
checking dev/speaker/speaker.h usability... no
checking dev/speaker/speaker.h presence... no
checking for dev/speaker/speaker.h... no
checking for mmap... no
checking for ioperm... no
checking for i386_iopl in -li386... no
checking for amd64_iopl in -lamd64... no
checking for windres... windres
checking for strip... strip
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /usr/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports the include directive... yes (GNU style)
checking whether make supports nested variables... yes
checking dependency style of gcc... gcc3
checking whether make supports nested variables... (cached) yes
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: creating man/Makefile
config.status: creating man/bash-completion/Makefile
config.status: creating opl/Makefile
config.status: creating opl/examples/Makefile
config.status: creating pcsound/Makefile
config.status: creating pkg/Makefile
config.status: creating pkg/config.make
config.status: creating pkg/osx/Info-gnustep.plist
config.status: creating pkg/osx/Info.plist
config.status: creating rpm.spec
config.status: creating data/Makefile
config.status: creating src/Makefile
config.status: creating src/doom.appdata.xml
config.status: creating src/doom.desktop
config.status: creating src/doom-screensaver.desktop
config.status: creating src/doom/Makefile
config.status: creating src/heretic.appdata.xml
config.status: creating src/heretic.desktop
config.status: creating src/heretic/Makefile
config.status: creating src/hexen.appdata.xml
config.status: creating src/hexen.desktop
config.status: creating src/hexen/Makefile
config.status: creating src/resource.rc
config.status: creating src/setup-res.rc
config.status: creating src/setup/Makefile
config.status: creating src/setup/setup.desktop
config.status: creating src/setup/setup-manifest.xml
config.status: creating src/strife.appdata.xml
config.status: creating src/strife.desktop
config.status: creating src/strife/Makefile
config.status: creating textscreen/Makefile
config.status: creating textscreen/examples/Makefile
config.status: creating config.h
config.status: executing depfiles commands
```
No errors. Let's run make again.

3) error:
```
<snip>
Making all in .
make[3]: Entering directory '/home/xxx/rude-doom/opl'
  CC       opl.o
  CC       opl_linux.o
  CC       opl_obsd.o
  CC       opl_queue.o
  CC       opl_sdl.o
  CC       opl_timer.o
  CC       opl_win32.o
opl_win32.c:33:10: fatal error: m_argv.h: No such file or directory
   33 | #include "m_argv.h"
      |          ^~~~~~~~~~
compilation terminated.
make[3]: *** [Makefile:415: opl_win32.o] Error 1
make[3]: Leaving directory '/home/admin/rude-doom/opl'
make[2]: *** [Makefile:435: all-recursive] Error 1
make[2]: Leaving directory '/home/xxx/rude-doom/opl'
make[1]: *** [Makefile:580: all-recursive] Error 1
make[1]: Leaving directory '/home/xxx/rude-doom'
make: *** [Makefile:431: all] Error 2

```
4) copy m_argv.h from ./src to ./opl

5) next error:
```
<snip>
Making all in opl
make[2]: Entering directory '/home/xxx/rude-doom/opl'
Making all in .
make[3]: Entering directory '/home/xxx/rude-doom/opl'
  CC       opl_win32.o
In file included from opl_win32.c:33:
m_argv.h:23:10: fatal error: doomtype.h: No such file or directory
   23 | #include "doomtype.h"
      |          ^~~~~~~~~~~~
compilation terminated.
make[3]: *** [Makefile:415: opl_win32.o] Error 1
make[3]: Leaving directory '/home/xxx/rude-doom/opl'

```

6) copy doomtype.h from ./src to ./opl

7) next error
```
<snip>
Making all in examples
make[3]: Entering directory '/home/xxx/rude-doom/opl/examples'
  CC       droplay.o
  CCLD     droplay.exe
D:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/9.2.0/../../../../i686-w64-mingw32/bin/ld.exe: ../libopl.a(opl_win32.o): in function `OPL_Win32_Init':
D:\msys64\home\xxx\rude-doom\opl/opl_win32.c:136: undefined reference to `M_CheckParm'
collect2.exe: error: ld returned 1 exit status

```

**Comment by drfrag666** [comment]
Sorry i haven't tried compiling under MSYS2 with the makefile, that most likely won't work. I've only compiled with Code::Blocks and VS 2017.
 You mean TortoiseGit right? On those errors you get you could try reverting "Don't try to load the ioperm.sys driver without the -ioperm parameter.", that's to avoid an useless error message. But i'd be surprised if it compiled after that but if it does i could change it.
 I'm still mostly offline so i'll try to post this from my phone. Are you trying to compile a 64 bit build? I'm using the included TDM-GCC, i only do 32 bit builds. You need to edit the project build options (search directories) with the SDL folders. That externlib folder is to avoid the need to compile them yourself, i copy everything to the SDL directories. So you need to get ALL the SDL development libraries for MinGW (SDL_net-devel-1.2.8-VC.zip, SDL_mixer-devel-1.2.12-VC.zip , SDL-devel-1.2.15-mingw32.tar.gz) and put there the folders where you extract them (relative paths). Mine are in C:\DEV\SDL and RUDE is in C:\DEV\RUDE. So just edit the project (as usual) and set the directories accordingly to match yours.
 When i merge with Choco 3 you should be able to use CMake to generate the projects.

**Comment by tpoppins** [comment]
The ioperm.sys commit is fine, I had the same problems before it. 

Thanks to the SDL dev libs you listed I was finally able to build the DOOM project in CodeBlocks using the MinGW GCC that comes with CodeBlocks. The whole workspace won't build because of an error related to the Hexen project (apparently m_misc.c is missing); after I closed the Hexen project (as well as the Heretic and the Strife ones) the workspace was rebuilt successfully,

I'll post a step-by-step tutorial in my next post. Perhaps you could incorporate it into your GitHub page and/or README.

**Comment by drfrag666** [comment]
Heretic also compiles. I don't build the workspace. The codeblocks projects were broken and i only fixed those.

**Comment by tpoppins** [comment]
I spent some time writing a detailed tutorial then decided to do it from scratch to confirm each step. More errors popping up randomly at every step! Now I see why Fraggle and Fabian ditched CodeBlocks -- it's terrible. 

After adding the Compiler and Linker search paths in projects' Build Option I happened to look at projects' Properties. There's a Build Options button there as well (what a bizarre UI decision). And when you check the Search dirs _there_ you will see that CodeBlocks automatically prepended the paths you entered with an extra "../"! 

I'm no big fan of CMake but the sooner you adopt it the better. It couldn't be any worse than CodeBlocks.
